addons/website_product_comparison/controllers/main.py

```diff
@@ -0,0 +1,102 @@
+# -*- coding: utf-8 -*-
+from odoo import http
+from odoo.http import request
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+import json
+
+
+class WebsiteSaleProductComparison(WebsiteSale):
+
+    def get_compute_currency_and_context(self):
+        pricelist_context = dict(request.env.context)
+        pricelist = False
+        if not pricelist_context.get('pricelist'):
+            pricelist = request.website.get_current_pricelist()
+            pricelist_context['pricelist'] = pricelist.id
+        else:
+            pricelist = request.env['product.pricelist'].browse(pricelist_context['pricelist'])
+
+        from_currency = request.env.user.company_id.currency_id
+        to_currency = pricelist.currency_id
+        compute_currency = lambda price: from_currency.compute(price, to_currency)
+        return compute_currency, pricelist_context
+
+    @http.route('/shop/compare/', type='http', auth="public", website=True)
+    def product_compare(self, **post):
+        values = {}
+        if post.get('products'):
+            product_ids = [int(i) for i in post.get('products').split(',') if i.isdigit()]
+            # use search to check read access on each record/ids
+            products = request.env['product.product'].search([('id', 'in', product_ids)]).exists()
+            values['products'] = products
+
+            res = {}
+            for num, product in enumerate(products):
+                for var in product.attribute_line_ids:
+                    cat_name = var.attribute_id.category_id.name
+                    att_name = var.attribute_id.name
+                    res.setdefault(cat_name, {})
+                    if not res[cat_name].get(att_name):
+                        res[cat_name][att_name] = [' - '] * len(products)
+                    val = product.attribute_value_ids.filtered(lambda x: x.attribute_id == var.attribute_id)
+                    res[cat_name][att_name][num] = val and val[0].name or ' - '
```

Comment:
if I understand correctly, you shall always have one and only one attribute to val, so this can be replaced by 
`res[cat_name][att_name][num] = val.name`